<a href="https://colab.research.google.com/github/rasecotineb/datascience/blob/master/datamining/Itemset_Pattern.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Frequent Itemset Pattern Mining

In [8]:
#Se instala la librería pymining si es necesario
#pip install pymining
import sys
!{sys.executable} -m pip install --user pymining

In [9]:
#setea para la sesion la ruta relativa donde se encuentra el modulo "prefixspan"
sys.path.insert(1, "/root/.local/lib/python3.6/site-packages")

#Caso a resolver


Como ya discutimos en clases teoricas, el pre-procesamiento de los datos es una etapa importante, la
cual consume mucho tiempo (ahora veremos por qué). Se cuenta con una pequeña
base de transacciones la cual representa la compra realizada por 50 clientes durante 1 mes. Dichos clientes,
solamente podían escoger entre 20 productos de un catálogo en línea, es decir, que tenemos solamente 20
items representados por sus códigos numéricos. Por ejemplo, el item 12 representa el producto lentes de sol.
Los datos se encuentran en el archivo ***Datos Compras.csv.***
En este notebook se procede con lo siguiente:
1. Importar el archivo csv en python mediante Pandas
2. Mostrar la información siguiente:
a) La transacción con mayor número de objetos comprados.
b) El promedio del número de items comprados en la base de transacciones.
c) El producto más vendido.
d) El producto menos vendido.
3. Transformar el dataframe de Pandas en un el formato que pueda ser leído por la librería pymining.
4. Extraer las reglas de asociación existentes en los datos. Se prueba con diferentes umbrales, por ejemplo,
puede ser muy específico (valores altos para el soporte minimal y para la confianza minimal).
5. Se utilizan diferentes valores en los umbrales para encontrar la cantidad “correcta” de patrones
6. Se implemente un algoritmo que, a partir del vector de reglas de asociación, selecciones los top-5 patrones.
Se pueden usar diferentes criterios, por ejemplo, aquellos que contienen la mayor cantidad de items o
aquellos que tienen una mayor confianza, o incluso, aquellos que contienen un producto en especial.


Recordar que se cuenta con 50 clientes que pueden escoger entre un catalogo de 20 items. En el dataframe se muestra cada una de las transacciones (T1..T50) con los respectivos códigos de items adquiridos

In [10]:
import pandas as pd
from pymining import itemmining
from pymining import assocrules

df = pd.read_csv('https://drive.google.com/uc?id=15S4lUzXkT0J6xZksZJTCtmdFPJIVbTRS&export=download', header=None)
df


0   1     2     3     4     5     6     7     8
0    T1   6   3.0   1.0   9.0   NaN   NaN   NaN   NaN
1    T2  13  17.0  10.0   2.0  16.0   4.0   NaN   NaN
2    T3  15   9.0   4.0   NaN   NaN   NaN   NaN   NaN
3    T4  11  18.0   NaN   NaN   NaN   NaN   NaN   NaN
4    T5  16   7.0  11.0  14.0  10.0   9.0  12.0   4.0
5    T6  16  20.0   2.0  15.0   NaN   NaN   NaN   NaN
6    T7   7  20.0  17.0   2.0  10.0  16.0   NaN   NaN
7    T8   6  20.0   2.0  18.0  11.0   7.0  19.0   NaN
8    T9   2   NaN   NaN   NaN   NaN   NaN   NaN   NaN
9   T10  12  16.0  14.0   5.0  13.0   NaN   NaN   NaN
10  T11  18  19.0   4.0   NaN   NaN   NaN   NaN   NaN
11  T12   7  13.0   3.0   4.0   8.0   1.0  20.0   2.0
12  T13  20  15.0  17.0   4.0   NaN   NaN   NaN   NaN
13  T14  10   3.0  14.0  15.0  12.0  19.0  17.0   NaN
14  T15   1  14.0  16.0   8.0  18.0   NaN   NaN   NaN
15  T16   1  13.0   3.0  18.0  11.0   NaN   NaN   NaN
16  T17   7   1.0  20.0   4.0   NaN   NaN   NaN   NaN
17  T18  11   1.0   7.0  20.0   NaN   NaN   NaN   NaN
18  T19   7   6.0  13.0   3.0   NaN   NaN   NaN   NaN
19  T20   4   1.0   8.0  20.0   1.0   NaN   NaN   NaN
20  T21   5   7.0  17.0  20.0  11.0  12.0   NaN   NaN
21  T22   4  11.0   1.0   8.0   NaN   NaN   NaN   NaN
22  T23   5   2.0   6.0  20.0  14.0   1.0  15.0   NaN
23  T24  13  18.0   6.0   NaN   NaN   NaN   NaN   NaN
24  T25  10   5.0  16.0   4.0  13.0   NaN   NaN   NaN
25  T26   2  11.0  13.0  19.0   2.0   NaN   NaN   NaN
26  T27  11   8.0   9.0  12.0   1.0  13.0  20.0   NaN
27  T28  13   0.0  16.0   3.0  15.0   NaN   NaN   NaN
28  T29  10   2.0   9.0   5.0  11.0   NaN   NaN   NaN
29  T30   2   6.0  16.0  12.0   8.0  11.0   3.0   NaN
30  T31  16  14.0  10.0   9.0   3.0  15.0   NaN   NaN
31  T32  12  20.0   8.0   1.0  12.0  13.0   3.0   NaN
32  T33  17  11.0   NaN   NaN   NaN   NaN   NaN   NaN
33  T34   6  13.0   5.0   2.0  20.0   8.0  19.0   NaN
34  T35   7  18.0   2.0  20.0  13.0   6.0   NaN   NaN
35  T36  13   4.0  18.0   7.0  13.0  20.0  10.0   NaN
36  T37  20   NaN   NaN   NaN   NaN   NaN   NaN   NaN
37  T38  14   NaN   NaN   NaN   NaN   NaN   NaN   NaN
38  T39  10  11.0   NaN   NaN   NaN   NaN   NaN   NaN
39  T40   4   8.0  12.0   1.0  19.0   NaN   NaN   NaN
40  T41   3   5.0  10.0   4.0  11.0  14.0  12.0  20.0
41  T42   3  11.0  17.0   8.0   9.0  20.0   1.0   NaN
42  T43  16  20.0   8.0   NaN   NaN   NaN   NaN   NaN
43  T44  10  20.0  18.0   NaN   NaN   NaN   NaN   NaN
44  T45   0   3.0   NaN   NaN   NaN   NaN   NaN   NaN
45  T46   2   4.0   NaN   NaN   NaN   NaN   NaN   NaN
46  T47  17   NaN   NaN   NaN   NaN   NaN   NaN   NaN
47  T48  18   1.0  11.0  20.0   NaN   NaN   NaN   NaN
48  T49   1   7.0   6.0  17.0  11.0   9.0  19.0  20.0
49  T50   2   8.0   5.0  17.0  18.0  11.0   NaN   NaN

Vemos que los clientes T5,T49,T12 y T41 son los clientes que han comprado una mayor variedad de productos

In [11]:
#Se obtienen las transacciones con el mayor número de items comprados
serie = df.groupby(by=0).count().sum(axis=1).sort_values(ascending= False)
serie

0
T5     8
T49    8
T12    8
T41    8
T23    7
T36    7
T27    7
T14    7
T32    7
T34    7
T30    7
T42    7
T8     7
T31    6
T21    6
T2     6
T35    6
T50    6
T7     6
T28    5
T26    5
T29    5
T25    5
T20    5
T40    5
T16    5
T15    5
T10    5
T19    4
T17    4
T22    4
T13    4
T18    4
T1     4
T48    4
T6     4
T24    3
T3     3
T43    3
T44    3
T11    3
T33    2
T39    2
T4     2
T45    2
T46    2
T37    1
T38    1
T47    1
T9     1
dtype: int64

El promedio de items comprados por cada cliente es menos de 5 (4.74)

In [12]:
#Promedio de numero de items por transaccion
promedio = serie.sum()/len(serie)
promedio


4.74

Se sabe que en cada transacción, las columnas de la 1 a la 8 indican los códigos de los items adquiridos. Así se tiene que los items 20, 11 y 1 son los más vendidos con 22, 18, 16 ocasiones, respectivamente

In [13]:
#El producto mas vendido. 
cols=[1,2,3,4,5,6,7,8]
suma = pd.Series()
for i in cols:
  res = df[i].value_counts()  
#Suma de series
  suma = suma.add(res, fill_value=0)
suma.sort_values(ascending=False)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  This is separate from the ipykernel package so we can avoid doing imports until


20.0    22.0
11.0    18.0
1.0     16.0
2.0     15.0
13.0    15.0
4.0     14.0
3.0     12.0
8.0     12.0
7.0     11.0
10.0    11.0
16.0    11.0
18.0    11.0
12.0    10.0
17.0    10.0
6.0      9.0
9.0      8.0
14.0     8.0
5.0      8.0
19.0     7.0
15.0     7.0
0.0      2.0
dtype: float64

En resumen, se obtiene que el más vendido es el **item 20** con 22 unidades y el  menos vendido es el **item 0** con 2 unidades

##Transformación de estructuras para el uso de ItemMining

In [14]:
#Se transforma el dataframe al formato de tupla/lista considerando las columnas que contiene los items comprados
cols = [1,2,3,4,5,6,7,8]
transacciones = tuple(df[cols].to_records(index=False).tolist())
#Es necesario eliminar los valores 'nan' para poder procesar posteriormente las reglas de asociacion
trans = []   #lista vacia

for i in range(0,len(transacciones)):
# Convierte a serie cada uno de los registros y elimina los 'nan'. Los añade a una lista de tuplas
  trans.append(tuple(pd.Series(transacciones[i]).dropna()))  
#Ahora convierte la lista de tuplas en una tupla de tuplas  
trans = tuple(trans)
trans

((6.0, 3.0, 1.0, 9.0),
 (13.0, 17.0, 10.0, 2.0, 16.0, 4.0),
 (15.0, 9.0, 4.0),
 (11.0, 18.0),
 (16.0, 7.0, 11.0, 14.0, 10.0, 9.0, 12.0, 4.0),
 (16.0, 20.0, 2.0, 15.0),
 (7.0, 20.0, 17.0, 2.0, 10.0, 16.0),
 (6.0, 20.0, 2.0, 18.0, 11.0, 7.0, 19.0),
 (2.0,),
 (12.0, 16.0, 14.0, 5.0, 13.0),
 (18.0, 19.0, 4.0),
 (7.0, 13.0, 3.0, 4.0, 8.0, 1.0, 20.0, 2.0),
 (20.0, 15.0, 17.0, 4.0),
 (10.0, 3.0, 14.0, 15.0, 12.0, 19.0, 17.0),
 (1.0, 14.0, 16.0, 8.0, 18.0),
 (1.0, 13.0, 3.0, 18.0, 11.0),
 (7.0, 1.0, 20.0, 4.0),
 (11.0, 1.0, 7.0, 20.0),
 (7.0, 6.0, 13.0, 3.0),
 (4.0, 1.0, 8.0, 20.0, 1.0),
 (5.0, 7.0, 17.0, 20.0, 11.0, 12.0),
 (4.0, 11.0, 1.0, 8.0),
 (5.0, 2.0, 6.0, 20.0, 14.0, 1.0, 15.0),
 (13.0, 18.0, 6.0),
 (10.0, 5.0, 16.0, 4.0, 13.0),
 (2.0, 11.0, 13.0, 19.0, 2.0),
 (11.0, 8.0, 9.0, 12.0, 1.0, 13.0, 20.0),
 (13.0, 0.0, 16.0, 3.0, 15.0),
 (10.0, 2.0, 9.0, 5.0, 11.0),
 (2.0, 6.0, 16.0, 12.0, 8.0, 11.0, 3.0),
 (16.0, 14.0, 10.0, 9.0, 3.0, 15.0),
 (12.0, 20.0, 8.0, 1.0, 12.0, 13.0, 3.0),
 (17.0

En el paso anterior tenemos la información en el formato necesario para aplicar datamining y encontrar *insights*. Para estos se van a identificar aquellos items o combinación de items que aparezcan con mayor frecuencia sobre la base de valor mínimo (min_support)

In [25]:
#Se extrae el conjunto de itemset frecuente
relim_input = itemmining.get_relim_input(trans)
#min_support es el "sigma" (soporte minimo absoluto) -> Los elementos del itemset deben aparecer al menos N veces en las transacciones
min_support = 5
item_sets = itemmining.relim(relim_input, min_support=min_support)

for itemset in item_sets:
  print (itemset)

frozenset({15.0})
frozenset({19.0})
frozenset({5.0})
frozenset({9.0})
frozenset({9.0, 11.0})
frozenset({14.0})
frozenset({6.0})
frozenset({2.0, 6.0})
frozenset({20.0, 6.0})
frozenset({12.0})
frozenset({11.0, 12.0})
frozenset({17.0})
frozenset({17.0, 11.0})
frozenset({17.0, 20.0})
frozenset({7.0})
frozenset({11.0, 7.0})
frozenset({20.0, 7.0})
frozenset({10.0})
frozenset({16.0, 10.0})
frozenset({10.0, 4.0})
frozenset({16.0})
frozenset({18.0})
frozenset({18.0, 11.0})
frozenset({18.0, 20.0})
frozenset({3.0})
frozenset({3.0, 13.0})
frozenset({1.0, 3.0})
frozenset({8.0})
frozenset({8.0, 1.0})
frozenset({8.0, 1.0, 20.0})
frozenset({8.0, 11.0})
frozenset({8.0, 20.0})
frozenset({2.0})
frozenset({2.0, 13.0})
frozenset({2.0, 11.0})
frozenset({2.0, 20.0})
frozenset({4.0})
frozenset({1.0, 4.0})
frozenset({4.0, 20.0})
frozenset({13.0})
frozenset({20.0, 13.0})
frozenset({1.0})
frozenset({1.0, 11.0})
frozenset({1.0, 11.0, 20.0})
frozenset({1.0, 20.0})
frozenset({11.0})
frozenset({11.0, 20.0})
frozense

In [27]:
#se construyen las reglas de asociacion con un nivel de confianza X
min_confidence = 0.7
rules = assocrules.mine_assoc_rules(item_sets, min_support=min_support, min_confidence=min_confidence)
for rule in rules:
  print (rule)

(frozenset({8.0, 20.0}), frozenset({1.0}), 5, 0.7142857142857143)
(frozenset({1.0, 11.0}), frozenset({20.0}), 5, 0.7142857142857143)
(frozenset({7.0}), frozenset({20.0}), 9, 0.8181818181818182)


La regla de asociación permite identificar patrones de consumo entre los diferentes *itemsets*. Para esto se define un unmbral de soporte mínimo y un intervalo de confianza.
Por ejemplo, en  este caso, se tiene que el **item 7** origina el consumo del **item 20** con una certeza de **81.82%**.
De igual manera, se puede puede concluir que el consumo de los **items 8 y 20** da como como resultado, la compra del **item 1** con una certeza de **71.43%**. Estos patrones se dan en un soporte mínimo de frecuencia de 5

In [28]:
#Se convierte el objeto lista que contiene la regla de asociacion a DataFrame
df_rule=pd.DataFrame(rules)
#Se procede a ordenar por el nivel de confianza y luego por la cantidad de incidencias que se cumple la regla de asociacion
df_rule.sort_values(ascending=False,by=[3,2])

0       1  2         3
2        (7.0)  (20.0)  9  0.818182
0  (8.0, 20.0)   (1.0)  5  0.714286
1  (1.0, 11.0)  (20.0)  5  0.714286